cc_utilities/redcap_sync.py

```diff
@@ -111,22 +121,199 @@ def set_external_id_column(df, external_id_col):
     For the given external_id_col, drop any rows with no value and
     copy to a new column named "external_id"
     """
+    df = df.copy()
     df = df.dropna(subset=[external_id_col])
-    df.loc[:, "external_id"] = df[external_id_col]
+    df["external_id"] = df[external_id_col]
     return df
 
 
-def split_complete_and_incomplete_records(cases_df):
+def get_external_ids_and_dobs(
+    external_ids, db_url, external_id_col, table_name="patient"
+):
+    """
+    Look up records by CDMS ID and DOB. This will be used to reject records
+    that do not match, to avoid overwriting existing patient records with
+    another patient's data.
+
+    Returns a list of matching rows, as dictionaries with external_id_col values.
+    """
+    # Load table
+    engine = create_engine(db_url)
+    meta = MetaData(bind=engine)
+    table = Table(table_name, meta, autoload=True, autoload_with=engine)
+
+    # Validate columns
+    column_names = [col.name for col in table.columns]
+    assert DOB_FIELD in column_names, f"{DOB_FIELD} not in {table_name} table"
+    assert (
+        external_id_col in column_names
+    ), f"{external_id_col} not in {table_name} table"
+
+    # Define the query
+    query = select(
+        [getattr(table.c, external_id_col), getattr(table.c, DOB_FIELD)]
+    ).where(
+        getattr(table.c, external_id_col).in_(external_ids),
+        getattr(table.c, DOB_FIELD).isnot(None),
+        getattr(table.c, DOB_FIELD) != "",
+    )
+
+    # Execute
+    conn = engine.connect()
+    try:
+        result = conn.execute(query)
+        return [dict(row) for row in result.fetchall()]
+    finally:
+        conn.close()
+
+
+def get_records_matching_id_and_dob(
+    df, external_id_col, cdms_patients_data, external_ids
+):
+    """
+    Given the data from the CDMS SQL Mirror with CDMS_IDs and DOBs,
+    accept the record for sync if the DOB from REDCap (in df) matches
+    the DOB in the CDMS data. Also accept the record if a CDMS_ID was not returned
```

Comment:
It's possible I'm not following this correctly, but I thought we did not want to sync records without a CDMS ID at all (i.e., neither to PCC nor back to REDCap). Instead, we'd ignore them until such time as they show up in the DB...

OR let me know if I've got my wires crossed! 🙂

Reply:
Ah, I think I'm the one who had a backwards understanding of this requirement! Oops. I'll update this.